### DISCUSSION TITLE
Getting started with Quarkus, reactive MySQL datasource, & Kotlin

### DISCUSSION THREAD

**Discussion by jehrenzweig-leagueapps** [quarkusio/quarkus#26165] [Q&A]
I'm pretty new to Quarkus, reactive programming, & Kotlin, and I'm having trouble figuring out how to get my first reactive SQL query working.  Here's the repository class I have right now, which I've written using the "_[Reactive SQL Clients](https://quarkus.io/guides/reactive-sql-clients)_" page's "_[Prepared Queries](https://quarkus.io/guides/reactive-sql-clients#prepared-queries)_" info:
```
import io.vertx.mutiny.mysqlclient.MySQLPool
import io.vertx.mutiny.sqlclient.Row
import io.vertx.mutiny.sqlclient.RowSet
import io.vertx.mutiny.sqlclient.Tuple

@ApplicationScoped
class PotatoTypeJdbcReadRepository(
	@ReactiveDataSource("test") private val testDataSource: MySQLPool
) : IPotatoTypeReadRepository {
	override suspend fun getPotatoTypeByName(name: String): PotatoTypeEntity? {
		// Return a specific potato type that matches the specified name.
		val something = this.testDataSource
			.preparedQuery("select * from PotatoTypes where NAME = ? order by NAME")
			.execute(Tuple.of(name))
			.onItem().transform(RowSet::iterator) 
			.onItem().transform(iterator -> iterator.hasNext() ? from(iterator.next()) : null)

		// Placeholder; the SQL result is what I want to return
		return PotatoTypeEntity()
	}
}
```

----

*[UPDATE @ 2022-06-17]* 
I've created a repository with a sample application to help demonstrate this behavior:

* https://github.com/jehrenzweig-leagueapps/quarkus-reactive-programming-20220617

The `PotatoTypeReadOnlyRepository` class' [getPotatoTypeByName()](https://github.com/jehrenzweig-leagueapps/quarkus-reactive-programming-20220617/blob/main/src/main/kotlin/org/example/dataaccess/PotatoTypeReadOnlyRepository.kt#L37-L71) method successfully submits a SQL query & returns a `PotatoType` object, which is then returned as the API response body.

However, if I call the `PotatoTypeReadOnlyRepository` class' [getPotatoTypeByNameReactive()](https://github.com/jehrenzweig-leagueapps/quarkus-reactive-programming-20220617/blob/main/src/main/kotlin/org/example/dataaccess/PotatoTypeReadOnlyRepository.kt#L73-L97) 
method, the API endpoint that calls it returns a HTTP 500 response, because I'm doing one or more things wrong.

The problem here is that I can't figure out to write the reactive equivalent of my non-reactive repository method.  Any help or pointers would be greatly appreciated!

[a comment by an automated account is not shown]

**Reply by evanchooly** [reply]
I'm spitballing here as I don't have a place to test this syntax out but what if you passed `RowSet<PotatoTypes>::iterator`?  I'm otherwise not terribly familiar with the reactive APIs.

**Reply by jehrenzweig-leagueapps** [reply]
@evanchooly I'm pretty sure I already tried that, and it didn't change the exception message in any way.  I've already commented out all of my reactive code for now, so I can't immediately check to be sure.

FWIW I figured creating a repository with example code would be more helpful in facilitating this discussion.  But I also thought I'd post before that repo was ready, on the off-chance somebody with a lot of Kotlin experience immediately knew what the answer was.  I'll get a sample repo ready tomorrow & post a link here.

**Reply by gsmet** [reply]
@jehrenzweig-leagueapps not sure exactly what you're trying to achieve but just so you know you can use Quarkus without reactive programming if you're more familiar with it. Reactive programming is a possibility but you can do everything imperative if you prefer.

**Reply by jehrenzweig-leagueapps** [reply]
@gsmet Yeah, I have database calls working without using `io.quarkus:quarkus-reactive-mysql-client` to interact with my MySQL database using a reactive datasource.  I may have to go with that for the time being, since reactive programming is so different & I'm having trouble figuring out how to achieve the same results I already get using an imperative approach.  Thought I'd give it a try, though.

@evanchooly I had a bit more time than I'd expected today, so I gave the reactive method another try and got it to compile.  Here's what it looks like now:
```
val uniPotatoTypeEntity: Uni<PotatoTypeEntity?> = this.testReactiveDataSource
	.preparedQuery("select * from PotatoTypes where NAME = ? order by NAME")
	.execute(io.vertx.mutiny.sqlclient.Tuple.of(name))
	.onItem().transform(io.vertx.mutiny.sqlclient.RowSet<io.vertx.mutiny.sqlclient.Row>::iterator)
	.onItem().transform { iterator ->
		if (iterator.hasNext()) {
			PotatoTypeEntityMapper().getPotatoTypeEntity(iterator.next())
		} else {
			null
		}
	}
```

What I didn't originally understand in the official docs was that the `from()` method was not a built-in Kotlin or Mutiny method, but a custom method that mapped the `RowSet` to a POJO class instance.  Once I replaced `from()` with my own `PotatoTypeEntityMapper().getPotatoTypeEntity()` method, it works.

Now I'm stuck on something completely new & different: how to extract the `PotatoTypeEntity?` object from my `Uni<PotatoTypeEntity?>` object.

**Reply by jehrenzweig-leagueapps** [reply]
I just updated my original post with a link to a GitHub repository containing a sample application:

* https://github.com/jehrenzweig-leagueapps/quarkus-reactive-programming-20220617

It uses Quarkus DevServices & Flyway to create two new databases, for the life of the running application.  One DB is exposed using a non-reactive datasource, the other using a reactive datasource.  The non-reactive repository method works properly, but the reactive repository method equivalent is broken -- I'm doing one or more things wrong.
